clusterctl/main_integration_test.go

```diff
@@ -0,0 +1,174 @@
+limitations under the License.
+*/
+
+package main_test
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"syscall"
+	"testing"
+)
+
+const (
+	clusterctlBinaryName = "clusterctl"
+	workingDirectoryName = "clusterctl"
+)
+
+// run these tests with the flag "-update" to update the values stored in all of the golden files
+var update = flag.Bool("update", false, "update .golden files")
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	err := setupPrerequisites()
+	if err != nil {
+		fmt.Printf("unable to setup prerequisites: %v\n", err)
+		os.Exit(1)
+	}
+	os.Exit(m.Run())
+}
+
+func TestEmptyAndInvalidArgs(t *testing.T) {
+	testCases := []struct {
+		name            string
+		args            []string
+		exitCode        int
+		fixtureFilename string
+	}{
+		{"no arguments", []string{}, 0, "no-args.golden"},
+		{"no arguments with invalid flag", []string{"--invalid-flag"}, 1, "no-args-invalid-flag.golden"},
+		{"create with no arguments", []string{"create"}, 0, "create-no-args.golden"},
+		{"create with no arguments with invalid flag", []string{"create", "--invalid-flag"}, 1, "create-no-args-invalid-flag.golden"},
+		{"create cluster with no arguments", []string{"create", "cluster"}, 1, "create-cluster-no-args.golden"},
+		{"create cluster with no arguments with invalid flag", []string{"create", "cluster", "--invalid-flag"}, 1, "create-cluster-no-args-invalid-flag.golden"},
+		{"delete with no arguments", []string{"delete"}, 0, "delete-no-args.golden"},
+		{"delete with no arguments with invalid flag", []string{"delete", "--invalid-flag"}, 1, "delete-no-args-invalid-flag.golden"},
+		{"delete cluster with no arguments", []string{"delete", "delete"}, 1, "delete-cluster-no-args.golden"},
+		{"delete cluster with no arguments with invalid flag", []string{"delete", "delete", "--invalid-flag"}, 1, "delete-cluster-no-args-invalid-flag.golden"},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			output, err := runClusterctl(t, tc.args...)
+			compareExitCode(t, err, tc.exitCode)
+			compareOutput(t, output, tc.fixtureFilename)
+		})
+	}
+}
+
+func compareExitCode(t *testing.T, err error, expectedExitCode int) {
+	if expectedExitCode == 0 {
+		if err != nil {
+			t.Errorf("expected no error, got: %v", err)
+		}
+	} else {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			ws := exitError.Sys().(syscall.WaitStatus)
+			actualExitCode := ws.ExitStatus()
+			if actualExitCode != expectedExitCode {
+				t.Errorf("unexpected exit code: want %v, got %v", expectedExitCode, actualExitCode)
+			}
+		} else {
+			t.Fatalf("unable to retrieve exit code from error: %v", err)
+		}
+	}
+}
+
+func compareOutput(t *testing.T, actualOutput string, goldenFileName string) {
+	t.Helper()
+	if *update {
+		writeFixture(t, goldenFileName, actualOutput)
+	}
+	expectedOutput := loadFixture(t, goldenFileName)
+	if actualOutput != expectedOutput {
+		// TODO: format this using a diff tool / library
```

Comment:
Create and link to a bug for this please.

Reply:
I think this is an argument for using something like testify's assert library which would probably handle this. Created the issue below to track:

https://github.com/kubernetes-sigs/cluster-api/issues/255